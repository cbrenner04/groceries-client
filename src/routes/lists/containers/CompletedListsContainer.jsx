import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { ListGroup } from 'react-bootstrap';
import PropTypes from 'prop-types';

import Alert from '../../../components/Alert';
import List from '../components/List';
import ConfirmModal from '../../../components/ConfirmModal';
import axios from '../../../utils/api';

function CompletedListsContainer(props) {
  const [completedLists, setCompletedLists] = useState(props.completedLists);
  const [errors, setErrors] = useState('');
  const [success, setSuccess] = useState('');
  const [listToDelete, setListToDelete] = useState('');
  const [showDeleteConfirm, setShowDeleteConfirm] = useState(false);

  const dismissAlert = () => {
    setSuccess('');
    setErrors('');
  };

  const failure = ({ response, request, message }) => {
    if (response) {
      if (response.status === 401) {
        props.history.push({
          pathname: '/users/sign_in',
          state: { errors: 'You must sign in' },
        });
      } else if ([403, 404].includes(response.status)) {
        setErrors('List not found');
      } else {
        const responseTextKeys = Object.keys(response.data);
        const responseErrors = responseTextKeys.map((key) => `${key} ${response.data[key]}`);
        setErrors(responseErrors.join(' and '));
      }
    } else if (request) {
      setErrors('Something went wrong');
    } else {
      setErrors(message);
    }
  };

  const handleRefresh = async (list) => {
    dismissAlert();
    try {
      await axios.post(`/lists/${list.id}/refresh_list`, {});
      const refreshedList = completedLists.find((completedList) => completedList.id === list.id);
      refreshedList.refreshed = true;
      setSuccess('Your list was successfully refreshed.');
    } catch (error) {
      failure(error);
    }
  };

  const handleDelete = (list) => {
    setListToDelete(list);
    setShowDeleteConfirm(true);
  };

  const handleDeleteConfirm = async () => {
    dismissAlert();
    setShowDeleteConfirm(false);
    try {
      await axios.delete(`/lists/${listToDelete.id}`);
      const lists = completedLists.filter((cl) => cl.id !== listToDelete.id);
      setCompletedLists(lists);
      setSuccess('Your list was successfully deleted.');
    } catch (error) {
      failure(error);
    }
  };

  return (
    <>
      <h1>Completed Lists</h1>
      <Alert errors={errors} success={success} handleDismiss={dismissAlert} />
      <div className="clearfix">
        <Link to="/lists" className="float-right">
          Back to lists
        </Link>
        <p className="float-left">Previously refreshed lists are marked with an asterisk (*).</p>
      </div>
      <ListGroup>
        {completedLists.map((list) => (
          <List
            userId={list.user_id}
            list={list}
            key={list.id}
            onListDeletion={handleDelete}
            completed={list.completed}
            onListRefresh={handleRefresh}
            accepted
          />
        ))}
      </ListGroup>
      <ConfirmModal
        action="delete"
        body="Are you sure you want to delete this list?"
        show={showDeleteConfirm}
        handleConfirm={() => handleDeleteConfirm()}
        handleClear={() => setShowDeleteConfirm(false)}
      />
    </>
  );
}

CompletedListsContainer.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func,
  }),
  completedLists: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      created_at: PropTypes.string.isRequired,
      completed: PropTypes.bool.isRequired,
      users_list_id: PropTypes.number,
      owner_id: PropTypes.number,
    }),
  ),
};

export default CompletedListsContainer;
