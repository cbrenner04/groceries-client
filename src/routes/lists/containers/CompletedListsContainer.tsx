import React, { useState } from 'react';
import { Link, useNavigate } from 'react-router';
import { toast } from 'react-toastify';

import TitlePopover from 'components/TitlePopover';
import { usePolling } from 'hooks';
import type { IList, TUserPermissions } from 'typings';

import AcceptedLists from '../components/AcceptedLists';
import { fetchCompletedLists } from '../../list/utils';

interface ICompletedListContainer {
  userId: string;
  completedLists: IList[];
  currentUserPermissions: TUserPermissions;
}

const CompletedListsContainer: React.FC<ICompletedListContainer> = (props): React.JSX.Element => {
  const [completedLists, setCompletedLists] = useState(props.completedLists);
  const [currentUserPermissions, setCurrentUserPermissions] = useState(props.currentUserPermissions);
  const navigate = useNavigate();

  usePolling(async () => {
    try {
      const lists = await fetchCompletedLists({ navigate });
      /* istanbul ignore else */
      if (lists) {
        const { completedLists: updatedCompletedLists, currentUserPermissions: updatedUserPerms } = lists;
        const isSameSet = (newSet: IList[] | TUserPermissions, oldSet: IList[] | TUserPermissions): boolean =>
          JSON.stringify(newSet) === JSON.stringify(oldSet);
        const completedSame = isSameSet(updatedCompletedLists, completedLists);
        const userPermsSame = isSameSet(updatedUserPerms, currentUserPermissions);
        if (!completedSame) {
          setCompletedLists(updatedCompletedLists);
        }
        if (!userPermsSame) {
          setCurrentUserPermissions(updatedUserPerms);
        }
      }
    } catch (err: unknown) {
      const errorMessage = 'You may not be connected to the internet. Please check your connection.';
      toast(`${errorMessage} Data may be incomplete and user actions may not persist.`, {
        type: 'error',
        autoClose: 5000,
      });
    }
  }, 10000);

  return (
    <React.Fragment>
      <div className="clearfix mb-3">
        <Link to="/lists" className="float-end">
          Back to lists
        </Link>
      </div>
      <AcceptedLists
        title={
          <TitlePopover
            title="Completed Lists"
            message={'Previously refreshed lists are marked with an asterisk (*).'}
          />
        }
        completed={true}
        fullList={true}
        userId={props.userId}
        completedLists={completedLists}
        setCompletedLists={setCompletedLists}
        currentUserPermissions={currentUserPermissions}
        setCurrentUserPermissions={setCurrentUserPermissions}
        setIncompleteLists={/* istanbul ignore next */ (): undefined => undefined}
        incompleteLists={[]}
      />
    </React.Fragment>
  );
};

export default CompletedListsContainer;
