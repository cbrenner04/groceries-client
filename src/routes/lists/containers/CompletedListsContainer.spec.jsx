import React from 'react';
import { act, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import CompletedListsContainer from './CompletedListsContainer';
import axios from '../../../utils/api';

describe('CompletedListsContainer', () => {
  let props;
  const renderCompletedListsContainer = (props) => {
    return render(
      <MemoryRouter>
        <CompletedListsContainer {...props} />
      </MemoryRouter>,
    );
  };

  beforeEach(() => {
    jest.useFakeTimers();
    props = {
      userId: 'id1',
      completedLists: [
        {
          id: 'id1',
          name: 'foo',
          type: 'GroceryList',
          created_at: new Date('05/28/2020').toISOString(),
          completed: true,
          users_list_id: 'id1',
          owner_id: 'id1',
          user_id: 'id1',
          refreshed: false,
        },
        {
          id: 'id2',
          name: 'bar',
          type: 'BookList',
          created_at: new Date('05/28/2020').toISOString(),
          completed: true,
          users_list_id: 'id2',
          owner_id: 'id1',
          user_id: 'id1',
          refreshed: false,
        },
        {
          id: 'id3',
          name: 'baz',
          type: 'BookList',
          created_at: new Date('05/28/2020').toISOString(),
          completed: true,
          users_list_id: 'id3',
          owner_id: 'id2',
          user_id: 'id1',
          refreshed: false,
        },
      ],
      currentUserPermissions: {
        id1: 'write',
        id2: 'write',
        id3: 'read',
      },
    };
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('renders', () => {
    const { container } = renderCompletedListsContainer(props);

    expect(container).toMatchSnapshot();
  });

  it('updates via polling when different data is returned', async () => {
    axios.get = jest
      .fn()
      .mockResolvedValueOnce({
        data: {
          current_user_id: 'id1',
          completed_lists: [
            {
              id: 'id1',
              users_list_id: 'id1',
              name: 'foo',
              user_id: 'id1',
              type: 'GroceryList',
              created_at: new Date('05/31/2020').toISOString(),
              completed: true,
              refreshed: false,
              owner_id: 'id1',
            },
          ],
          current_list_permissions: { id1: 'write' },
        },
      })
      .mockResolvedValueOnce({
        data: {
          current_user_id: 'id1',
          completed_lists: [
            {
              id: 'id1',
              users_list_id: 'id1',
              name: 'foo',
              user_id: 'id1',
              type: 'GroceryList',
              created_at: new Date('05/31/2020').toISOString(),
              completed: true,
              refreshed: false,
              owner_id: 'id1',
            },
            {
              id: 'id2',
              users_list_id: 'id2',
              name: 'bar',
              user_id: 'id1',
              type: 'GroceryList',
              created_at: new Date('05/31/2020').toISOString(),
              completed: true,
              refreshed: false,
              owner_id: 'id1',
            },
          ],
          current_list_permissions: { id1: 'write' },
        },
      });

    const { getByTestId, queryByTestId } = renderCompletedListsContainer(props);

    await act(async () => {
      jest.advanceTimersByTime(10000);
    });

    expect(axios.get).toHaveBeenCalledTimes(1);
    expect(getByTestId('list-id1')).toBeVisible();
    expect(queryByTestId('list-id2')).toBeNull();

    await act(async () => {
      jest.advanceTimersByTime(10000);
    });

    expect(axios.get).toHaveBeenCalledTimes(2);
    expect(getByTestId('list-id1')).toBeVisible();
    expect(getByTestId('list-id2')).toBeVisible();
  });

  it('does not update via polling when different data is not returned', async () => {
    axios.get = jest.fn().mockResolvedValue({
      data: {
        current_user_id: 'id1',
        completed_lists: [
          {
            id: 'id1',
            users_list_id: 'id1',
            name: 'foo',
            user_id: 'id1',
            type: 'GroceryList',
            created_at: new Date('05/31/2020').toISOString(),
            completed: true,
            refreshed: false,
            owner_id: 'id1',
          },
        ],
        current_list_permissions: { id1: 'write' },
      },
    });

    const { getByTestId } = renderCompletedListsContainer(props);

    await act(async () => {
      jest.advanceTimersByTime(10000);
    });

    expect(axios.get).toHaveBeenCalledTimes(1);
    expect(getByTestId('list-id1')).toBeVisible();

    await act(async () => {
      jest.advanceTimersByTime(10000);
    });

    expect(axios.get).toHaveBeenCalledTimes(2);
    expect(getByTestId('list-id1')).toBeVisible();
  });
});
