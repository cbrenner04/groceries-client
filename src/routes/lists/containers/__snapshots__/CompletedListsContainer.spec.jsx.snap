// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompletedListsContainer renders 1`] = `
<div>
  <div
    class="clearfix mb-3"
  >
    <a
      class="float-end"
      href="/lists"
    >
      Back to lists
    </a>
  </div>
  <div
    class="mb-4"
  >
    <div
      class="clearfix"
    >
      <div
        class="mx-auto float-start"
      >
        <div
          class="m-0 row"
        >
          <h2
            class="text-capitalize pe-0 w-auto"
          >
            Completed Lists
          </h2>
          <i
            class="far fa-question-circle float-end text-secondary ps-0 w-auto"
            data-test-id="Completed Lists-popover"
          />
        </div>
      </div>
      <button
        class="mx-auto float-end btn btn-link"
        type="button"
      >

         Select
      </button>
    </div>
    <div
      class="list-group"
    >
      <div
        class="accepted-list list-list-group-item list-group-item"
        data-test-class="completed-list"
        data-test-id="list-id1"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-12 col-12"
          >
            <div
              class=" pt-1 row"
            >
              <div
                class="col-lg-6"
              >
                <a
                  class="router-link"
                  href="/lists/id1"
                >
                  <h5
                    class="mb-1"
                  >
                    <i
                      class="fa fa-shopping-bag text-secondary me-3"
                    />
                    foo
                  </h5>
                </a>
              </div>
              <div
                class="col-lg-4"
              >
                <small
                  class="text-muted"
                >
                  May 28 2020, 12:00:00 am
                </small>
              </div>
              <div
                class=" pe-lg-3 col-lg-2"
              >
                <div
                  class="float-end btn-group"
                  role="group"
                >
                  <button
                    class="p-0 me-3 btn btn-link"
                    data-test-id="complete-list-refresh"
                    type="button"
                  >
                    <i
                      class="fa fa-redo fa-2x text-primary"
                    />
                  </button>
                  <button
                    class="p-0 btn btn-link"
                    data-test-id="complete-list-trash"
                    type="button"
                  >
                    <i
                      class="fa fa-trash fa-2x text-danger"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="accepted-list list-list-group-item list-group-item"
        data-test-class="completed-list"
        data-test-id="list-id2"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-12 col-12"
          >
            <div
              class=" pt-1 row"
            >
              <div
                class="col-lg-6"
              >
                <a
                  class="router-link"
                  href="/lists/id2"
                >
                  <h5
                    class="mb-1"
                  >
                    <i
                      class="fa fa-book text-secondary me-3"
                    />
                    bar
                  </h5>
                </a>
              </div>
              <div
                class="col-lg-4"
              >
                <small
                  class="text-muted"
                >
                  May 28 2020, 12:00:00 am
                </small>
              </div>
              <div
                class=" pe-lg-3 col-lg-2"
              >
                <div
                  class="float-end btn-group"
                  role="group"
                >
                  <button
                    class="p-0 me-3 btn btn-link"
                    data-test-id="complete-list-refresh"
                    type="button"
                  >
                    <i
                      class="fa fa-redo fa-2x text-primary"
                    />
                  </button>
                  <button
                    class="p-0 btn btn-link"
                    data-test-id="complete-list-trash"
                    type="button"
                  >
                    <i
                      class="fa fa-trash fa-2x text-danger"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="accepted-list list-list-group-item list-group-item"
        data-test-class="completed-list"
        data-test-id="list-id3"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-12 col-12"
          >
            <div
              class=" pt-1 row"
            >
              <div
                class="col-lg-6"
              >
                <a
                  class="router-link"
                  href="/lists/id3"
                >
                  <h5
                    class="mb-1"
                  >
                    <i
                      class="fa fa-book text-secondary me-3"
                    />
                    baz
                  </h5>
                </a>
              </div>
              <div
                class="col-lg-4"
              >
                <small
                  class="text-muted"
                >
                  May 28 2020, 12:00:00 am
                </small>
              </div>
              <div
                class=" pe-lg-3 col-lg-2"
              >
                <div
                  class="float-end btn-group"
                  role="group"
                >
                  <button
                    class="p-0 me-3 btn btn-link"
                    data-test-id="complete-list-refresh"
                    disabled=""
                    type="button"
                  >
                    <i
                      class="fa fa-redo fa-2x text-primary"
                    />
                  </button>
                  <button
                    class="p-0 btn btn-link"
                    data-test-id="complete-list-trash"
                    type="button"
                  >
                    <i
                      class="fa fa-trash fa-2x text-danger"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
